src/main/java/com/puppycrawl/tools/checkstyle/xpath/ElementNode.java

```diff
@@ -75,12 +78,30 @@ public ElementNode(AbstractNode root, AbstractNode parent, DetailAST detailAst)
         this.parent = parent;
         this.root = root;
         this.detailAst = detailAst;
+        nodeDepth = -1;
```

Comment:
@timurt @rnveach I didn't know whether to touch other modules or not. It can be done and will be more efficient. So, shall I create a `depth` attribute of `DetailASTImpl` class and pass on `this.depth + 1` while creating child nodes? Let me know what should be done, meanwhile I'll add the tests

Reply:
@rnveach @gaurabdg 
No, I am saying about creating `depth` attribute for `AbstractNode` class, 
So my suggestion is to initialize `depth` attribute during Xpath tree creation process instead of lazy calculations. 

For example constructor of the `ElementNode` could be smth like
```
 public ElementNode(AbstractNode root, AbstractNode parent, DetailAST detailAst) {
        super(root.getTreeInfo());
        this.parent = parent;
        this.root = root;
        this.detailAst = detailAst;
        this.nodeDepth = parent.nodeDepth + 1; // <-- You can use parent's height
        text = TokenUtil.getTokenName(detailAst.getType());
        indexAmongSiblings = parent.getChildren().size();
        createTextAttribute();
        createChildren();
    }
```

Xpath tree construction is recursive DFS, so height of every child of the node is larger by one than the parent height

What do you think about it?

Reply:
@timurt I understand what you are saying and also agree. But the tree contruction is happening and being returned as `DetailAST` right? So, `the depth+1` passing needs to happen there.
Also, while recursively browsing the code for tree generation I see the `RootNode` is being passed on to here, at `XPathUtil`
```
    /**
     * Returns xpath query results on file as string.
     *
     * @param xpath query to evaluate
     * @param file file to run on
     * @return all results as string separated by delimiter
     * @throws CheckstyleException if some parsing error happens
     * @throws IOException if an error occurs
     */
    public static String printXpathBranch(String xpath, File file) throws CheckstyleException,
            IOException {
        final XPathEvaluator xpathEvaluator = new XPathEvaluator();
        try {
            final RootNode rootNode = new RootNode(JavaParser.parseFile(file,
                JavaParser.Options.WITH_COMMENTS));
            final XPathExpression xpathExpression = xpathEvaluator.createExpression(xpath);
            final XPathDynamicContext xpathDynamicContext =
                xpathExpression.createDynamicContext(rootNode);
            final List<Item<?>> matchingItems = xpathExpression.evaluate(xpathDynamicContext);
            return matchingItems.stream()
                .map(item -> ((AbstractNode) item).getUnderlyingNode())
                .map(AstTreeStringPrinter::printBranch)
                .collect(Collectors.joining(DELIMITER));
        }
        catch (XPathException ex) {
            final String errMsg = String.format(Locale.ROOT,
                "Error during evaluation for xpath: %s, file: %s", xpath, file.getCanonicalPath());
            throw new CheckstyleException(errMsg, ex);
        }
    }
```
So, is the tree is gettng created by the saxon libraries? It would be really helpful if you could tell me about the tree creation flow.

Reply:
@gaurabdg 
We have two trees:
- one is AST tree with `DetailAST` objects, used by almost all checks to validate rules
- another one is Xpath tree with classes implementing `NodeInfo`(*saxon*) interface, this tree is used only for querying nodes using xpath queries

Xpath tree nodes have reference to corresponding AST tree nodes

Construction of the xpath tree starts with `RootNode`, as argument `RootNode` constructor takes root `DetailAst` object, and recursively traverse AST tree and creates corresponding Xpath nodes

For example for `SuppressionXpathFilter`, we create `RootNode` [here](https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/filters/XpathFilterElement.java#L235), 
Violation event [TreeWalkerAuditEvent](https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/TreeWalkerAuditEvent.java#L49) itself contains reference to `rootAst`, so we can create Xpath tree from it

is it clear? I know it is not obvious, so please ask questions

Reply:
@timurt Thanks a lot for clarifying. I will ask as and when required while implementing the new approach.